src/batch/src/executor/join/hash_join.rs

```diff
@@ -12,319 +12,1403 @@
+/// case, we use `ChunkedData` to link them together. For example:
+///
+/// | id | key | row |
+/// | --- | --- | --- |
+/// | 0 | 1 | (1, 2, 3) |
+/// | 1 | 4 | (4, 5, 6) |
+/// | 2 | 1 | (1, 3, 7) |
+/// | 3 | 1 | (1, 3, 2) |
+/// | 4 | 3 | (3, 2, 1) |
+///
+/// The corresponding join hash map is:
+///
+/// | key | value |
+/// | --- | --- |
+/// | 1 | 0 |
+/// | 4 | 1 |
+/// | 3 | 4 |
+///
+/// And we save build rows with the same key like this:
+///
+/// | id | value |
+/// | --- | --- |
+/// | 0 | 2 |
+/// | 1 | None |
+/// | 2 | 3 |
+/// | 3 | None |
+/// | 4 | None |
+///
+/// This can be seen as an implicit linked list. For convenience, we use `RowIdIter` to iterate all
+/// build side row ids with the given key.
+type JoinHashMap<K> = HashMap<K, RowId, PrecomputedBuildHasher>;
+
+struct RowIdIter<'a> {
+    current_row_id: Option<RowId>,
+    next_row_id: &'a ChunkedData<Option<RowId>>,
+}
+
+impl ChunkedData<Option<RowId>> {
+    fn row_id_iter(&self, begin: Option<RowId>) -> RowIdIter {
+        RowIdIter {
+            current_row_id: begin,
+            next_row_id: self,
+        }
     }
+}
+
+impl<'a> Iterator for RowIdIter<'a> {
+    type Item = RowId;
 
-    #[inline(always)]
-    pub(super) fn right_len(&self) -> usize {
-        self.right_col_len
+    fn next(&mut self) -> Option<Self::Item> {
+        self.current_row_id.map(|row_id| {
+            self.current_row_id = self.next_row_id[row_id];
+            row_id
+        })
     }
+}
 
-    #[inline(always)]
-    pub(super) fn build_key_columns(&self) -> &[usize] {
-        &self.right_key_columns
+struct EquiJoinParams<K> {
+    probe_side: BoxedExecutor,
+    probe_data_types: Vec<DataType>,
+    probe_key_idxs: Vec<usize>,
+    build_side: Vec<DataChunk>,
+    build_data_types: Vec<DataType>,
+    full_data_types: Vec<DataType>,
+    hash_map: JoinHashMap<K>,
+    next_build_row_with_same_key: ChunkedData<Option<RowId>>,
+}
+
+#[derive(Default)]
+struct LeftNonEquiJoinParams {
+    probe_column_count: usize,
+    first_output_row_id: Vec<usize>,
+    has_more_output_rows: bool,
+    found_matched: bool,
+}
+
+#[derive(Default)]
+struct RightNonEquiJoinParams {
+    build_row_ids: Vec<RowId>,
+    build_row_matched: ChunkedData<bool>,
+}
+
+impl<K: HashKey> HashJoinExecutor<K> {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(self: Box<Self>) {
+        let probe_data_types = self.probe_side_source.schema().data_types();
+        let build_data_types = self.build_side_source.schema().data_types();
+        let full_data_types = [probe_data_types.clone(), build_data_types.clone()].concat();
+
+        let mut build_side = Vec::new();
+        let mut build_row_count = 0;
+        #[for_await]
+        for build_chunk in self.build_side_source.execute() {
+            let build_chunk = build_chunk?;
+            // Assume build_chunk is compacted already.
```

Comment:
We can't have assumption here, either do compaction or do assertion here. I think it's better to do compaction here.